File: src/pages/api/generate-schedule.js
import prisma from '../../lib/db';
import fs from 'fs/promises';

const daysOfWeek = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
const timeSlots = [
    '08:30', '09:00', '09:30', '10:00', '10:30', '11:00', '11:30', '12:00',
    '12:30', '13:00', '13:30', '14:00', '14:30', '15:00', '15:30', '16:00',
    '16:30', '17:00', '17:30', '18:00', '18:30', '19:00', '19:30', '20:00',
    '20:30', '21:00', '21:30', '22:00', '22:30'
];

// Debugging helper
const logDebug = (message, data = null) => {
    console.log(`[DEBUG] ${message}`);
    if (data !== null) {
        console.log(JSON.stringify(data, null, 2));
    }
};

// Fetch unique room names and truncate them
async function fetchRooms() {
    try {
        logDebug('Fetching unique room names from database...');
        const rooms = await prisma.class.findMany({
            select: { Room: true },
            distinct: ['Room'],
            orderBy: { Room: 'asc' }
        });
        const roomNames = rooms.map(({ Room }) => Room.replace(/-.*$/, ''));
        logDebug('Fetched and truncated room names:', roomNames);
        return roomNames;
    } catch (error) {
        console.error('Error fetching rooms:', error);
        return [];
    }
}

// Check room availability for a specific slot
async function checkRoomAvailability(room, day, startTime, endTime) {
    try {
        logDebug(`Checking availability for room ${room} on ${day} from ${startTime} to ${endTime}...`);
        const result = await prisma.class.findFirst({
            where: {
                Room: room,
                Day: day,
                NOT: {
                    OR: [
                        { EndTime: { lte: startTime } },
                        { StartTime: { gte: endTime } }
                    ]
                }
            }
        });
        const isAvailable = result.length === 0;
        logDebug(`Result: ${JSON.stringify(result, null, 2)}`);
        logDebug(`Room ${room} availability: ${isAvailable ? 'Available' : 'Not Available'}`);
        return isAvailable;
    } catch (error) {
        console.error(`Error checking availability for room ${room}:`, error);
        return false;
    }
}

// Generate the schedule data
async function generateScheduleData() {
    logDebug('Starting schedule data generation...');
    const roomNames = await fetchRooms();
    const schedule = [];

    for (const day of daysOfWeek) {
        const dayData = { day, rooms: [] };
        logDebug(`Processing day: ${day}`);

        for (const room of roomNames) {
            const roomData = { room, availability: [] };
            logDebug(`Processing room: ${room}`);

            for (let i = 0; i < timeSlots.length - 1; i++) {

                // Start time is current time slot + 1 minute. Replace the last character of each time slot with '1'
                const startTime = timeSlots[i].slice(0, -1) + '1';
                const endTime = startTime.slice(0, -1) + '2';
                const available = await checkRoomAvailability(room, day, startTime, endTime);
                roomData.availability.push(available ? 1 : 0);
            }

            dayData.rooms.push(roomData);
        }

        schedule.push(dayData);
    }

    logDebug('Completed schedule data generation.');
    return schedule;
}

// Save the generated schedule data to a file
async function saveScheduleData() {
    try {
        logDebug('Saving schedule data to JSON file...');
        const scheduleData = await generateScheduleData();
        await fs.writeFile('./scheduleData.json', JSON.stringify(scheduleData, null, 2));
        logDebug('Schedule data saved successfully.');
    } catch (error) {
        console.error('Error saving schedule data:', error);
    }
}

// Export for API usage
export default async function handler(req, res) {
    try {
        logDebug('Handler triggered via API request.');
        await saveScheduleData();
        res.status(200).json({ message: 'Schedule generated successfully' });
    } catch (error) {
        console.error('Error:', error);
        res.status(500).json({ error: 'Failed to generate schedule' });
    }
}

File: generate_schedule.py
import asyncio
import json
import os
import asyncpg
from pathlib import Path

# Define constants
DAYS_OF_WEEK = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
TIME_SLOTS = [
    '08:30', '09:00', '09:30', '10:00', '10:30', '11:00', '11:30', '12:00',
    '12:30', '13:00', '13:30', '14:00', '14:30', '15:00', '15:30', '16:00',
    '16:30', '17:00', '17:30', '18:00', '18:30', '19:00', '19:30', '20:00',
    '20:30', '21:00', '21:30', '22:00', '22:30'
]

DB_URL = os.getenv("DATABASE_URL")
print(DB_URL)
input()

async def fetch_rooms(connection):
    print("Fetching unique room names from database...")
    query = """
        SELECT DISTINCT "Room"
        FROM "classes"
        ORDER BY "Room" ASC;
    """
    rows = await connection.fetch(query)
    rooms = [row['Room'] for row in rows]
    rooms = [room.split('-')[0] for room in rooms]
    print(f"Rooms fetched: {rooms}")
    return rooms

async def check_room_availability(connection, room, day, start_time, end_time):
    print(f"Checking availability for room {room} on {day} from {start_time} to {end_time}...")

    query = """
        SELECT 1
        FROM "classes"
        WHERE "Room" LIKE $1
          AND "Day" = $2
          AND NOT ("EndTime" <= $3 OR "StartTime" >= $4)
        LIMIT 1;
    """    
    result = await connection.fetchrow(query, f"{room}%", day, start_time, end_time)
    available = result is None
    print(f"Room {room} availability: {'Available' if available else 'Not Available'}")
    return available

async def generate_schedule_data():
    print("Starting schedule data generation...")
    
    # Await the coroutine to get the connection object
    connection = await asyncpg.connect(DB_URL)  
    
    try:
        room_names = await fetch_rooms(connection)
        schedule = []

        for day in DAYS_OF_WEEK:
            print(f"Processing day: {day}")
            day_data = {"day": day, "rooms": []}

            for room in room_names:
                room_data = {"room": room, "availability": []}

                for i in range(len(TIME_SLOTS) - 1):

                    start_time = TIME_SLOTS[i][:-1] + '1'
                    end_time = TIME_SLOTS[i][:-1] + '2'
                    available = await check_room_availability(connection, room, day, start_time, end_time)
                    room_data["availability"].append(1 if available else 0)

                day_data["rooms"].append(room_data)

            schedule.append(day_data)

        return schedule
    finally:
        await connection.close()  # Always close the connection


async def save_schedule_data():
    print("Saving schedule data to JSON file...")
    schedule_data = await generate_schedule_data()
    output_path = Path("./scheduleData.json")
    with output_path.open("w") as file:
        json.dump(schedule_data, file, indent=2)
    print(f"Schedule data saved to {output_path.resolve()}")

if __name__ == "__main__":
    asyncio.run(save_schedule_data())
